data_steward/utils/bq.py

```diff
@@ -193,6 +193,50 @@ def define_dataset(project_id, dataset_id, description, label_or_tag):
     return dataset
 
 
+def update_labels_and_tags(dataset_id,
+                           existing_labels_or_tags,
+                           new_labels_or_tags,
+                           overwrite_ok=False):
+    """
+    Updates labels or tags in dataset if not set or needing to be updated
+    or overwrites existing labels or tags in the dataset
+
+    :param dataset_id:  string name to identify the dataset
+    :param existing_labels_or_tags:  labels already existing on the dataset = Dict[str, str]
+                                     tags already existing on the dataset = Dict[str, '']
+    :param new_labels_or_tags:  new labels to add to the dataset = Dict[str, str]
+                                new tags to add to the dataset = Dict[str, '']
+    :param overwrite_ok:  flag to signal if labels or tags are to be either
+                             overwritten (False as default) or updated (True)
+
+    :raises:  RuntimeError if parameters are not specified
+    :raises:  RuntimeError if overwrite_ok is false and the new labels and tags provided
+                already exist in the existing labels or tags dictionary
+
+    :return:  a dictionary of new labels or tags
+    """
+    if not dataset_id:
+        raise RuntimeError("Provide a dataset_id")
+
+    if not new_labels_or_tags:
+        raise RuntimeError("Please provide a label or tag")
+
+    if overwrite_ok is False and any(new_labels_or_tags.items() &
+                                     existing_labels_or_tags.items()) is True:
+        raise RuntimeError(
+            "Cannot overwrite. The provided new label or tag already exists")
+
+    if overwrite_ok is True and any(new_labels_or_tags.items() &
+                                    existing_labels_or_tags.items()) is False:
+        labels_and_tags = dict(
+            list(existing_labels_or_tags.items()) +
+            list(new_labels_or_tags.items()))
+        return labels_and_tags
+
+    else:
+        return new_labels_or_tags
```

Comment:
Okay, just so I'm understanding your snippet of code:

- Updates should occur even if the new labels/tags supplied already exist
- An error is thrown if existing labels/tags are supplied with the overwrite_ok flag being False
- The existing labels/tags and the updated labels/tags (existing + new) will be returned

Am I understanding that correctly?

Reply:
* Yes. Remember our function here does not actually do updates, it just gives us the final set of labels/tags the dataset should have.
* Yes, an error should be thrown if `overwrite_ok==False`, but specifically only for _labels_ where the value is being changed (the condition doesn't make sense for _tags_). So if  someone tries to do the following it would fail.
    ```python
    bq.update_labels(dataset_id, { 'label_1': 'apples', 'tag_1': ''}, { 'label_1': 'oranges', 'tag_1': ''}, overwrite_ok=False)
    ```
* Yes, it returns the final set of labels.
Please let me know if anything else is unclear.